### PR TITLE
Add poll.vote action and tests.

### PR DESCRIPTION
**Pull request by reiterl**

### CODE REVIEW

File: tests/system/action/poll/test_vote.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+        response = self.client.post(
+            "/",
+            json=[
+                {
+                    "action": "poll.vote",
+                    "data": [
+                        {
+                            "id": 1,
+                            "user_id": 1,
+                            "value": {"11": "Y", "12": "N", "13": "A"},
```
Comment (jsangmeister) [inline tests/system/action/poll/test_vote.py:98]: This should be invalid: In a `YN` poll, you cannot give `A` as a value.

File: tests/system/action/poll/test_vote.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+            json=[
+                {
+                    "action": "poll.vote",
+                    "data": [{"id": 1, "user_id": 1, "value": "X"}],
+                }
+            ],
+        )
+        self.assert_status_code(response, 400)
+        assert (
+            "data.value must be valid by one of anyOf definition"
```
Comment (jsangmeister) [inline tests/system/action/poll/test_vote.py:213]: This is not a good error to send to the client. Unfortunately, jsonschema isn't very helpful there. I would suggest that you don't try to mix the schematas like you currently do, but instead leave the schema of the action itself empty and first check the pollmethod of the poll and then manually check with the correct schema. This way, you don't have to use `anyOf` (hopefully) and can give reasonable error messages.

File: tests/system/action/poll/test_vote.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+            "/",
+            json=[
+                {
+                    "action": "poll.vote",
+                    "data": [{"id": 1, "user_id": 1, "value": "Y"}],
+                }
+            ],
+        )
+        self.assert_status_code(response, 400)
+        assert "poll.vote is not allowed for analog voting." in response.data.decode()
```
Comment (jsangmeister) [inline tests/system/action/poll/test_vote.py:235]: See above, this should be allowed

**Review by jsangmeister — CHANGES_REQUESTED** [review]
@reiterl Did you think about a way to port the tests from OS3? There are quite a lot of them and they all test things that still need to be tested here. I don't know If we want to put everything in one PR (will probably get really big). Maybe we open another voting branch, like in OS3, @FinnStutzenstein?

I have just looked roughly through this, but there are still a lot of features missing. This can probably best be tested by first porting the tests and then working until they are all passing.

File: openslides_backend/action/actions/poll/vote.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        gmr = GetManyRequest(
+            Collection("group"),
+            group_ids,
+            ["user_ids"],
+        )
+        result = self.datastore.get_many([gmr])
+        db_groups = result.get(Collection("group"), {})
+        for group in db_groups:
+            if user_id in db_groups[group].get("user_ids", []):
+                return
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/poll/vote.py:102]: Update the check mechanism here.
use user's groups of the meeting to compare to entitled_group_ids.

File: openslides_backend/action/actions/poll/vote.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        ]
+        self.execute_other_action(OptionSetAutoFields, payload)
+
+
+def check_value_for_option_vote(value: Union[str, Dict[str, Any]]) -> bool:
+    return isinstance(value, dict)
+
+
+def check_value_for_global_vote(value: Union[str, Dict[str, Any]]) -> bool:
+    return isinstance(value, str)
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/poll/vote.py:220]: I replaced them.

File: openslides_backend/action/actions/poll/vote.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        user_id: int,
+        payload: List[Dict[str, Any]],
+    ) -> None:
+        for key in value:
+            weight = "1.000000"
+            used_value = value[key]
+
+            if self.poll["pollmethod"] in ("Y", "N"):
+                weight = "1.000000" if value[key] == 1 else "0.000000"
+                used_value = self.poll["pollmethod"]
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/poll/vote.py:137]: Use the `vote_weight`.

File: tests/system/action/poll/test_vote.py (hunk 3 above)
Comment (reiterl) [reply tests/system/action/poll/test_vote.py:235]: This is correct, we have the option.update action for updating analog polls.

File: openslides_backend/action/actions/poll/vote.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+        # check for double vote
+        if user_id in self.poll.get("voted_ids", []):
+            raise ActionException("Only one vote per poll per user allowed.")
+        instance["voted_ids"] = self.poll.get("voted_ids", [])
+        instance["voted_ids"].append(user_id)
+
+        # check for analog type
+        if self.poll.get("type") == "analog":
+            raise ActionException("poll.vote is not allowed for analog voting.")
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/poll/vote.py:59]: This got changed. We want to use the option.update function for updating the options.

File: openslides_backend/action/actions/poll/vote.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                            "anyOf": [
+                                {"type": "integer"},
+                                {"type": "string", "enum": ["Y", "N", "A"]},
+                            ]
+                        },
+                    },
+                ]
+            },
+        },
+    )
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/poll/vote.py]: Replaced the AnyOf, with code validation.

File: tests/system/action/poll/test_vote.py (hunk 2 above)
Comment (reiterl) [reply tests/system/action/poll/test_vote.py]: Replaced AnyOf with code validation.

File: tests/system/action/poll/test_vote.py (hunk 1 above)
Comment (reiterl) [reply tests/system/action/poll/test_vote.py:119]: Add error raising for this.

**Review by jsangmeister — APPROVED** [review]
Can be merged for now. We need to add the OS3 tests ASAP to fix possible errors.

**Comment by jsangmeister** [comment]
rebase needed.

**Comment by reiterl** [comment]
Rebased, and used the new vote_weight_$ field.

**Pull request merged**